### PR TITLE
[MB-2492] Create MTO Support Service Object

### PR DESCRIPTION
**Pull request by Ronolibert**
## Description

Refactor support endpoints create MTO to be a service object in a separate support service objects directory

## Reviewer Notes
- Ran into an issue with the payload formatting where I can't use the internal payload functions since those are private so I ended up reusing those exact functions in the service object code
- I also ended up adding reference ID default behavior so we don't have to make up a reference ID each time
## Setup
Create a data.json in the root file like so but with your respective IDs:
```
{
  "body": {
    "contractorID": "5db13bb4-6d29-4bdb-bc81-262f4513ecf6",
    "moveOrder": {
      "customer": {
        "firstName": "Grace",
        "lastName": "Griffin",
        "agency": "MARINES",
        "email": "ggriffin@example.com"
      },
      "entitlement": {
        "nonTemporaryStorage": false,
        "totalDependents": 91
      },
      "destinationDutyStationID": "425ca6bc-0e41-4d98-ba5f-239fa2e1b3de",
      "originDutyStationID": "b768b7c5-aa0a-4043-b0e7-430688b8663c"
    }
  }
}
```
`make server_run`
`prime-api-client support-create-move-task-order --insecure --filename data.json  | jq`
You should see the MTO payload

## Code Review Verification Steps
* [ ] Request review from a member of a different team.
* [ ] Have the Jira acceptance criteria been met for this change?

## References

* [Jira story](https://dp3.atlassian.net/browse/MB-2492) for this change

### CODE REVIEW

File: pkg/services/support/move_task_order/move_task_order_creator.go [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+	} else if err != nil {
+		logger.Error("supportapi.createOrGetCustomer error", zap.Error(err))
+		e := services.NewQueryError("Customer", err, "Unable to create Customer")
+		return nil, e
+	}
+	return customer, nil
+
+}
+
+// CustomerModel converts payload to model - currently does not tackle addresses
```
Comment (shimonacarvalho) [inline pkg/services/support/move_task_order/move_task_order_creator.go:181]: Wondering why these payload to model functions were moved here vs. the payload_to_model files?
Seems like we wouldn't want duplicates of these, but to use the shared payload_to_model versions.

Comment (sandy-wright) [reply pkg/services/support/move_task_order/move_task_order_creator.go:181]: It looks like Ron had to do this because of the permissions issue with the payload_to_model versions. I would suggest doing the conversions before using the service function to create the MTO, ex:

```go
// in CreateMoveTaskOrderHandler:
moveTaskOrder := payloads.MoveTaskOrderModel(params.Body)
moveOrder := payloads.MoveOrderModel(params.Body.MoveOrder)  // customer and entitlement models are created here

moveTaskOrder, err := h.moveTaskOrderCreator.InternalCreateMoveTaskOrder(moveTaskOrder, moveOrder, logger)
```

This would mean refactoring a good chunk of what's in the service object now, but I _think_ it should work well and maybe help streamline some of the audits? One of the issues is that we have redundant ID fields, one on the base object definition, and one in the subobject definition (so we have `moveOrder.customerID` and `moveOrder.customer.ID`). I think we should consider eliminating this redundancy and just use the ID in the subobject definition, but I would like to hear what others think about this. If we're okay with eliminating this redundancy in the request, the refactor should work! 

This is a lot of work to request though, so I'm not super sure how we should handle it for this PR 😬. We could open these other issues in new tickets, maybe?

File: pkg/services/support/move_task_order/move_task_order_creator.go [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+func (f moveTaskOrderCreator) InternalCreateMoveTaskOrder(payload supportmessages.MoveTaskOrder, logger handlers.Logger) (*models.MoveTaskOrder, error) {
+	var moveTaskOrder *models.MoveTaskOrder
+	var refID string
+	if payload.MoveOrder == nil {
+		return nil, services.NewQueryError("MoveTaskOrder", nil, "MoveOrder is necessary")
+	}
+
+	transactionError := f.db.Transaction(func(tx *pop.Connection) error {
+		// Create or get customer
+		customer, err := createOrGetCustomer(tx, customer.NewCustomerFetcher(f.db), payload.MoveOrder.CustomerID.String(), payload.MoveOrder.Customer, logger)
```
[the remark this answers is not in the document]
Comment (Ronolibert) [reply pkg/services/support/move_task_order/move_task_order_creator.go:36]: I think it should be using the tx incase it fails so we don't continue with the rest of the tx

**Comment by Ronolibert** [comment]
@sandy-wright We could add that to a new ticket! I was contemplating doing it in this ticket but the scope of it was already larger than I expected when I was working on it

**Comment by shimonacarvalho** [comment]
Can you summarize the status Ron? 
Link the new ticket for the remaining work, let us know if all open comments are resolved?

File: pkg/services/support/move_task_order/move_task_order_creator.go (hunk 1 above)
Comment (shimonacarvalho) [reply pkg/services/support/move_task_order/move_task_order_creator.go:181]: What's the permissions issue @sandy-wright ? I'm not understanding that bit.

And with the redundant ID fields, this is an example of where there is some business logic behind it. If customerID is provided, the endpoint expects that the customer already exists. If no customerID is provided, endpoint expects it should create the customer. In fact I don't think customer.ID is included? Although I do think it /should/ be returned in the response.

Comment (sandy-wright) [reply pkg/services/support/move_task_order/move_task_order_creator.go:181]: I think it's something about importing from private packages in Golang - and the `payload_to_model` code is in a package that private to the support API. I don't really understand how this works, full disclaimer 😬 

I think we've made a purpose for this field using that logic, but that we could do the same thing with the ID in the customer payload. I don't think there's any real benefit in having two fields for the same piece of data because the backend can handle that logic either way. And I totally agree that the customer ID should be returned in the response!

**Comment by sandy-wright** [comment]
I think the missed transaction `tx` ref on line 36 of `pkg/services/support/move_task_order/move_task_order_creator.go` needs to be fixed, but after that, I'd be happy to approve this and create a ticket or two to pick up the other work we pointed out.

File: pkg/services/support/move_task_order/move_task_order_creator.go (hunk 1 above)
Comment (shimonacarvalho) [reply pkg/services/support/move_task_order/move_task_order_creator.go:181]: Hmm, I can see that, actually. 

Although I kind of feel like we should make that change after we do the adr thing. Or else it may unnecessarily churn back to the original implementation after discussion/feedback.

File: pkg/services/support/move_task_order/move_task_order_creator.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (shimonacarvalho) [reply pkg/services/support/move_task_order/move_task_order_creator.go:36]: Ron maybe you misunderstood sandy's comment? in the params of `createOrGetCustomer` you create a `customer.NewCustomerFetcher`, but pass in f.db to that, I think that's the one she's referring to.

[the remark this answers is not in the document]
Comment (Ronolibert) [reply pkg/services/support/move_task_order/move_task_order_creator.go:36]: You're right! Totally missed that, changed the service object to use the transaction connection instead

**Comment by Ronolibert** [comment]
@sandy-wright Sorry I misunderstood! Changed it to use the missing transaction

@shimonacarvalho I'm not sure if I'm fully capturing the task, but I made this card https://dp3.atlassian.net/browse/MB-2952
